### PR TITLE
[Multi-task Learning] Add multi-task Dataloader for multi-task learning

### PR DESCRIPTION
**Pull request by classicsong**
*Issue #, if available:*
#789 

*Description of changes:*
Add GSgnnMultiTaskDataLoader to support multi-task learning. 

When initializing a GSgnnMultiTaskDataLoader, users need to provide two inputs: 1) a list of config.TaskInfo objects recording the information of each task and 2) a list of dataloaders corresponding to each training task.  

During training for each iteration, GSgnnMultiTaskDataLoader will iteratively call each task-dataloader to generate a mini-batch and finally return a list of mini-batches to the trainer.

The length of the dataloader (number of batches for an epoch) is determined by the largest task in the GSgnnMultiTaskDataLoader.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tests/unit-tests/test_dataloading.py [hunk 1]
```diff
@@ -30,10 +30,17 @@
     generate_dummy_dist_graph,
     generate_dummy_dist_graph_reconstruct,
     generate_dummy_dist_graph_homogeneous_failure_graph,
+    generate_dummy_dist_graph_multi_task,
     create_distill_data,
+    SIZE_DICT
 )
 
 import graphstorm as gs
+from graphstorm.config import (TaskInfo,
```
Comment (zhjwy9343) [inline tests/unit-tests/test_dataloading.py:39]: Should we import EdgeRegression as well?

File: tests/unit-tests/test_dataloading.py [hunk 2]
```diff
@@ -2181,7 +2189,215 @@ def test_GSgnnTrainData_homogeneous():
...
+        len_edge2_0 = len(task_edge_2_dataloader)
+        assert len_edge2_0 == 5
+
+        assert len(multi_dataloader) == 10
+
+        len(multi_dataloader.dataloaders) == 7
+        dataloaders = multi_dataloader.dataloaders
+        len(multi_dataloader.task_infos) == 7
+
+        dataloaders = multi_dataloader.dataloaders
```
Comment (zhjwy9343) [inline tests/unit-tests/test_dataloading.py:2316]: Duplicated line as the #2313.

File: python/graphstorm/dataloading/dataloading.py [hunk 3]
```diff
@@ -1686,6 +1686,111 @@ def __len__(self):
...
+    def __init__(self, dataset, task_infos, task_dataloaders):
+        assert len(task_infos) == len(task_dataloaders), \
+            "Number of task_info should match number of task dataloaders"
+        # check dataloaders
+        lens = []
+        for task_info, dataloader in zip(task_infos, task_dataloaders):
+            assert isinstance(dataloader, (GSgnnEdgeDataLoaderBase,
+                                           GSgnnLinkPredictionDataLoaderBase,
+                                           GSgnnNodeDataLoaderBase)), \
+                "The task data loader should be a GSgnnEdgeDataLoaderBase " \
```
[the remark this answers is not in the document]
Comment (classicsong) [reply python/graphstorm/dataloading/dataloading.py:1712]: I changed it into an instance of xxx

File: tests/unit-tests/test_dataloading.py (hunk 1 above)
Comment (classicsong) [reply tests/unit-tests/test_dataloading.py:39]: EdgeRegression is not used.

**Review by zhjwy9343 — APPROVED** [review]
LGTM

**Pull request merged**